File: apps-microservices/api-gateway/app/routers/sso.py
"""OAuth 2.1 + PKCE client for account-service SSO.

Adds /auth/login and /auth/callback to api-gateway. Sets the same
request.session["user"] dict the rest of the app already understands,
so existing DocsAuthMiddleware + /login handler keep working as fallback.

Env vars (defaults match the docker-compose wiring):
    ACCOUNT_BASE_URL       — public URL of account-service (e.g. http://account-service-backend:8600 in-cluster, https://account.hellopro.fr in prod)
    ACCOUNT_REDIRECT_URI   — e.g. http://localhost:8050/auth/callback (must match the URI registered in account-service)

Client credentials are resolved by `common_utils.sso.get_account_credentials()`,
which derives env keys from `SERVICE_NAME` (e.g. SERVICE_NAME=api-gateway →
ACCOUNT_CLIENT_ID_API_GATEWAY + ACCOUNT_CLIENT_SECRET_API_GATEWAY) and falls
back to plain ACCOUNT_CLIENT_ID + ACCOUNT_CLIENT_SECRET when those aren't set.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import logging
import os
import secrets
import time
from typing import Optional

import httpx
from fastapi import APIRouter, HTTPException, Request, Response
from fastapi.responses import RedirectResponse

from common_utils.sso import (
    AccountCredentialsMissing,
    get_account_credentials,
    get_account_credentials_from_api,
)

logger = logging.getLogger("sso")

router = APIRouter(tags=["SSO"])

ACCOUNT_BASE_URL = os.environ.get("ACCOUNT_BASE_URL", "http://account-service-backend:8600")
ACCOUNT_REDIRECT_URI = os.environ.get("ACCOUNT_REDIRECT_URI", "")

# Credentials are resolved lazily on first /auth/login request: env first
# (instant, no network), then HTTP fallback to /internal/credentials/{name}
# on account-service. Cache once we get them so we don't refetch per request.
_cached_credentials: Optional[tuple[str, str]] = None


async def _get_credentials() -> tuple[str, str]:
    global _cached_credentials
    if _cached_credentials:
        return _cached_credentials
    try:
        _cached_credentials = get_account_credentials()
        return _cached_credentials
    except AccountCredentialsMissing:
        pass
    _cached_credentials = await get_account_credentials_from_api()
    return _cached_credentials

REPLAY_WINDOW_S = 5 * 60


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


@router.get("/auth/login", include_in_schema=False)
async def auth_login() -> Response:
    """Start the PKCE flow: 302 to account-service /authorize with a fresh challenge."""
    if not ACCOUNT_REDIRECT_URI:
        raise HTTPException(500, "ACCOUNT_REDIRECT_URI not configured")
    try:
        client_id, _ = await _get_credentials()
    except AccountCredentialsMissing as exc:
        raise HTTPException(500, f"account-service credentials unavailable: {exc}")

    verifier = _b64url(secrets.token_bytes(32))
    challenge = _b64url(hashlib.sha256(verifier.encode()).digest())
    state = _b64url(secrets.token_bytes(16))

    target = (
        f"{ACCOUNT_BASE_URL}/authorize"
        f"?response_type=code"
        f"&client_id={client_id}"
        f"&redirect_uri={ACCOUNT_REDIRECT_URI}"
        f"&code_challenge={challenge}"
        f"&code_challenge_method=S256"
        f"&state={state}"
    )
    response = RedirectResponse(target, status_code=302)
    secure_cookie = os.environ.get("SECURE_COOKIE", "false").lower() in {"1", "true", "yes"}
    response.set_cookie("auth_verifier", verifier, httponly=True, samesite="lax", secure=secure_cookie, max_age=600, path="/")
    response.set_cookie("auth_state", state, httponly=True, samesite="lax", secure=secure_cookie, max_age=600, path="/")
    return response


@router.get("/auth/callback", include_in_schema=False)
async def auth_callback(request: Request, code: Optional[str] = None, state: Optional[str] = None) -> Response:
    """Exchange the authorization code for tokens, then populate request.session."""
    if not code or not state:
        raise HTTPException(400, "missing code or state")

    stored_state = request.cookies.get("auth_state")
    verifier = request.cookies.get("auth_verifier")
    if stored_state != state:
        raise HTTPException(400, "state mismatch")
    if not verifier:
        raise HTTPException(400, "missing verifier")

    try:
        client_id, client_secret = await _get_credentials()
    except AccountCredentialsMissing as exc:
        raise HTTPException(500, f"account-service credentials unavailable: {exc}")

    async with httpx.AsyncClient(timeout=10) as cli:
        r = await cli.post(
            f"{ACCOUNT_BASE_URL}/token",
            data={
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": ACCOUNT_REDIRECT_URI,
                "code_verifier": verifier,
            },
            auth=(client_id, client_secret),
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

    if r.status_code != 200:
        logger.warning("token exchange failed status=%s body=%s", r.status_code, r.text)
        raise HTTPException(502, f"token exchange failed: {r.text}")

    tokens = r.json()
    access_token = tokens["access_token"]
    refresh_token = tokens.get("refresh_token", "")

    payload_segment = access_token.split(".")[1]
    payload_segment += "=" * (-len(payload_segment) % 4)
    import json as _json
    claims = _json.loads(base64.urlsafe_b64decode(payload_segment))

    request.session["user"] = {
        "display_name": claims.get("name") or claims.get("sub"),
        "email": claims.get("sub") or claims.get("email"),
        "token": access_token,
        "sso": {
            "sid": claims.get("sid"),
            "iss": claims.get("iss"),
            "exp": claims.get("exp"),
            "refresh_token": refresh_token,
        },
    }

    response = RedirectResponse("/docs", status_code=303)
    response.delete_cookie("auth_verifier", path="/")
    response.delete_cookie("auth_state", path="/")
    return response


@router.post("/auth/logout-webhook", include_in_schema=False)
async def logout_webhook(request: Request) -> Response:
    """Account-service back-channel logout. Drops the local session if the
    HMAC matches and the iat is within the replay window."""
    try:
        _, client_secret = await _get_credentials()
    except AccountCredentialsMissing as exc:
        raise HTTPException(500, f"account-service credentials unavailable: {exc}")

    body = await request.body()
    presented = request.headers.get("X-Logout-Signature", "")
    expected = "sha256=" + hmac.new(client_secret.encode(), body, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(presented, expected):
        raise HTTPException(401, "bad signature")

    import json as _json
    try:
        evt = _json.loads(body)
    except Exception:
        raise HTTPException(400, "bad body")

    iat = int(evt.get("iat", 0))
    if abs(time.time() - iat) > REPLAY_WINDOW_S:
        raise HTTPException(401, "stale event")

    sub = evt.get("sub")
    sid = evt.get("sid")
    logger.info("[sso] back-channel logout received sub=%s sid=%s (no-op without server-side session store)", sub, sid)

    return Response(status_code=204)

File: apps-microservices/api-gateway-go/internal/routers/sso.go
package routers

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gin-contrib/sessions"
	"github.com/gin-gonic/gin"

	"api-gateway-go/internal/sso"
)

// SSODeps holds all dependencies for the SSO route handlers.
type SSODeps struct {
	Resolver        *sso.Resolver
	AccountBaseURL  string // internal base URL for token exchange (server-to-server)
	AccountPubURL   string // public-facing base URL for browser redirects
	AccountRedirect string // redirect_uri sent to the authorization server
	SecureCookie    bool
	HTTPClient      *http.Client
}

// replayWindowS is the maximum allowed age (in seconds) of a logout webhook event.
const replayWindowS = 5 * 60

// RegisterSSO registers /auth/login, /auth/callback, and /auth/logout-webhook on r.
func RegisterSSO(r *gin.Engine, d SSODeps) {
	if d.HTTPClient == nil {
		d.HTTPClient = &http.Client{Timeout: 10 * time.Second}
	}
	r.GET("/auth/login", authLoginHandler(d))
	r.GET("/auth/callback", authCallbackHandler(d))
	r.POST("/auth/logout-webhook", logoutWebhookHandler(d))
}

// redirectURI returns the configured redirect URI or falls back to the first
// registered URI returned by the account service.
func redirectURI(d SSODeps, c *sso.ClientCredentials) (string, error) {
	if d.AccountRedirect != "" {
		return d.AccountRedirect, nil
	}
	if len(c.RedirectURIs) > 0 {
		return c.RedirectURIs[0], nil
	}
	return "", fmt.Errorf("no redirect_uri available (set ACCOUNT_REDIRECT_URI or register one)")
}

// authLoginHandler initiates the PKCE authorization code flow.
// Stores the verifier and state in short-lived HttpOnly cookies.
func authLoginHandler(d SSODeps) gin.HandlerFunc {
	return func(c *gin.Context) {
		creds, err := d.Resolver.Resolve(c.Request.Context())
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"detail": fmt.Sprintf("account-service credentials unavailable: %v", err)})
			return
		}
		ru, err := redirectURI(d, creds)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
			return
		}
		p := sso.NewPKCEPair()
		target := fmt.Sprintf(
			"%s/authorize?response_type=code&client_id=%s&redirect_uri=%s&code_challenge=%s&code_challenge_method=S256&state=%s",
			d.AccountPubURL,
			url.QueryEscape(creds.ClientID),
			url.QueryEscape(ru),
			p.Challenge,
			p.State,
		)
		setShortCookie(c, "auth_verifier", p.Verifier, d.SecureCookie)
		setShortCookie(c, "auth_state", p.State, d.SecureCookie)
		c.Redirect(http.StatusFound, target)
	}
}

// setShortCookie sets a short-lived (10 min) HttpOnly cookie for PKCE state.
func setShortCookie(c *gin.Context, name, value string, secure bool) {
	c.SetSameSite(http.SameSiteLaxMode)
	c.SetCookie(name, value, 600, "/", "", secure, true)
}

// authCallbackHandler exchanges the authorization code for tokens, validates
// the PKCE state, stores user info in the session, and redirects to /docs.
func authCallbackHandler(d SSODeps) gin.HandlerFunc {
	return func(c *gin.Context) {
		code := c.Query("code")
		state := c.Query("state")
		if code == "" || state == "" {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "missing code or state"})
			return
		}
		storedState, _ := c.Cookie("auth_state")
		verifier, _ := c.Cookie("auth_verifier")
		if storedState != state {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "state mismatch"})
			return
		}
		if verifier == "" {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "missing verifier"})
			return
		}
		creds, err := d.Resolver.Resolve(c.Request.Context())
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"detail": fmt.Sprintf("account-service credentials unavailable: %v", err)})
			return
		}
		ru, err := redirectURI(d, creds)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
			return
		}

		form := url.Values{}
		form.Set("grant_type", "authorization_code")
		form.Set("code", code)
		form.Set("redirect_uri", ru)
		form.Set("code_verifier", verifier)

		req, _ := http.NewRequestWithContext(c.Request.Context(), "POST", d.AccountBaseURL+"/token", strings.NewReader(form.Encode()))
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
		req.SetBasicAuth(creds.ClientID, creds.ClientSecret)
		resp, err := d.HTTPClient.Do(req)
		if err != nil {
			c.JSON(http.StatusBadGateway, gin.H{"detail": "token exchange failed: " + err.Error()})
			return
		}
		defer resp.Body.Close()
		body, _ := io.ReadAll(resp.Body)
		if resp.StatusCode != http.StatusOK {
			c.JSON(http.StatusBadGateway, gin.H{"detail": "token exchange failed: " + string(body)})
			return
		}
		var tokens map[string]any
		if err := json.Unmarshal(body, &tokens); err != nil {
			c.JSON(http.StatusBadGateway, gin.H{"detail": "token exchange parse failed"})
			return
		}
		access, _ := tokens["access_token"].(string)
		refresh, _ := tokens["refresh_token"].(string)
		claims := decodeJWTPayload(access)

		s := sessions.Default(c)
		s.Set("user", map[string]any{
			"display_name": firstNonEmpty(claims["name"], claims["sub"]),
			"email":        firstNonEmpty(claims["sub"], claims["email"]),
			"token":        access,
			"sso": map[string]any{
				"sid":           claims["sid"],
				"iss":           claims["iss"],
				"exp":           claims["exp"],
				"refresh_token": refresh,
			},
		})
		_ = s.Save()

		// Clear PKCE cookies now that flow is complete.
		c.SetCookie("auth_verifier", "", -1, "/", "", d.SecureCookie, true)
		c.SetCookie("auth_state", "", -1, "/", "", d.SecureCookie, true)
		c.Redirect(http.StatusSeeOther, "/docs")
	}
}

// decodeJWTPayload decodes the claims from the payload segment of a JWT
// without verifying the signature. Used only for display/session values;
// authoritative validation is done server-side by the account service.
func decodeJWTPayload(token string) map[string]any {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return map[string]any{}
	}
	seg := parts[1]
	if pad := len(seg) % 4; pad > 0 {
		seg += strings.Repeat("=", 4-pad)
	}
	raw, err := base64.URLEncoding.DecodeString(seg)
	if err != nil {
		return map[string]any{}
	}
	var out map[string]any
	_ = json.Unmarshal(raw, &out)
	return out
}

// firstNonEmpty returns the first non-empty string (or non-nil value) from values.
func firstNonEmpty(values ...any) any {
	for _, v := range values {
		if s, ok := v.(string); ok && s != "" {
			return s
		}
		if v != nil {
			return v
		}
	}
	return nil
}

// logoutWebhookHandler handles back-channel logout events from the account service.
// Verifies the HMAC-SHA256 signature and rejects stale events outside the replay window.
func logoutWebhookHandler(d SSODeps) gin.HandlerFunc {
	return func(c *gin.Context) {
		creds, err := d.Resolver.Resolve(c.Request.Context())
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"detail": fmt.Sprintf("account-service credentials unavailable: %v", err)})
			return
		}
		body, err := io.ReadAll(c.Request.Body)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "bad body"})
			return
		}
		mac := hmac.New(sha256.New, []byte(creds.ClientSecret))
		mac.Write(body)
		expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))
		presented := c.GetHeader("X-Logout-Signature")
		if !hmac.Equal([]byte(expected), []byte(presented)) {
			c.JSON(http.StatusUnauthorized, gin.H{"detail": "bad signature"})
			return
		}
		var evt map[string]any
		if err := json.Unmarshal(body, &evt); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "bad body"})
			return
		}
		iat, _ := evt["iat"].(float64)
		if absInt64(time.Now().Unix()-int64(iat)) > replayWindowS {
			c.JSON(http.StatusUnauthorized, gin.H{"detail": "stale event"})
			return
		}
		_ = context.Background() // placeholder for future session invalidation by sid
		c.Status(http.StatusNoContent)
	}
}

func absInt64(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}
